File: sdk/go/release_context.go
package faas

import (
	"context"
	"net/http"
	"strings"
)

// GregaleRevisionHeader identifies an exact deployment for one app. It is
// deliberately not propagated to downstream services: revision pins are
// scoped to the app that received them.
const GregaleRevisionHeader = "X-Gregale-Revision"

// GregaleReleaseHeader identifies an immutable project deployment graph.
const GregaleReleaseHeader = "X-Gregale-Release"

type gregaleReleaseContextKey struct{}

// WithGregaleRelease returns a context carrying the selected project release.
// The value is treated as an opaque ID; an empty or malformed value is ignored.
func WithGregaleRelease(ctx context.Context, release string) context.Context {
	release = cleanGregaleRelease(release)
	if release == "" {
		return ctx
	}
	return context.WithValue(ctx, gregaleReleaseContextKey{}, release)
}

// GregaleReleaseFromContext returns the release captured for the current
// request, if any.
func GregaleReleaseFromContext(ctx context.Context) (string, bool) {
	release, ok := ctx.Value(gregaleReleaseContextKey{}).(string)
	return release, ok && release != ""
}

// GregaleReleaseMiddleware captures the release selected by Gregale on an
// inbound HTTP request and makes it available to outbound service calls made
// with that request's context.
//
// The gateway validates the incoming release before forwarding it. This
// middleware preserves that context; it does not grant authority to select a
// deployment on its own.
func GregaleReleaseMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		values := r.Header.Values(GregaleReleaseHeader)
		if len(values) == 1 {
			if release := cleanGregaleRelease(values[0]); release != "" {
				r = r.WithContext(WithGregaleRelease(r.Context(), release))
			}
		}
		next.ServeHTTP(w, r)
	})
}

// NewGregaleReleaseTransport returns a RoundTripper that forwards the
// request-scoped release to managed Gregale service hosts (for example,
// billing.svc.gregale). It never forwards a revision pin to a service because
// that pin belongs to the caller app, not the target app. Requests to other
// hosts are left untouched.
func NewGregaleReleaseTransport(base http.RoundTripper) http.RoundTripper {
	if base == nil {
		base = http.DefaultTransport
	}
	return gregaleReleaseRoundTripper{base: base}
}

type gregaleReleaseRoundTripper struct {
	base http.RoundTripper
}

func (t gregaleReleaseRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	if req.URL == nil || !isGregaleServiceHost(req.URL.Hostname()) {
		return t.base.RoundTrip(req)
	}

	// Work on a shallow request/header copy to honor RoundTripper's contract
	// not to mutate the caller's request.
	cloned := req.Clone(req.Context())
	cloned.Header = req.Header.Clone()
	cloned.Header.Del(GregaleRevisionHeader)

	if release, ok := GregaleReleaseFromContext(req.Context()); ok &&
		cleanGregaleRelease(cloned.Header.Get(GregaleReleaseHeader)) == "" {
		cloned.Header.Set(GregaleReleaseHeader, release)
	}
	return t.base.RoundTrip(cloned)
}

func isGregaleServiceHost(host string) bool {
	host = strings.ToLower(strings.TrimSuffix(host, "."))
	return strings.HasSuffix(host, ".svc.gregale")
}

func cleanGregaleRelease(release string) string {
	release = strings.TrimSpace(release)
	if release == "" || strings.ContainsAny(release, ", \t\r\n") {
		return ""
	}
	return release
}

File: sdk/node/src/release-context.ts
import { AsyncLocalStorage } from 'node:async_hooks';

/** Header carrying a deployment-scoped pin for a public app request. */
export const GREGALE_REVISION_HEADER = 'X-Gregale-Revision';

/** Header carrying an immutable project deployment graph. */
export const GREGALE_RELEASE_HEADER = 'X-Gregale-Release';

const releaseContext = new AsyncLocalStorage<string | undefined>();

function cleanRelease(value: string | null | undefined): string | undefined {
  const release = value?.trim();
  return release && !/[\s,]/.test(release) ? release : undefined;
}

/**
 * Run a request handler in the release context selected by Gregale. Pass the
 * inbound request's release header; the context is isolated across concurrent
 * async handlers.
 *
 * Revision pins are intentionally not captured here: they are scoped to the
 * app that received them and must not follow an internal service call.
 */
export function withGregaleReleaseContext<T>(release: string | null | undefined, handler: () => T): T {
  return releaseContext.run(cleanRelease(release), handler);
}

/** Capture the header from a Web Request or Headers object and run a handler. */
export function withGregaleRequestContext<T>(headers: HeadersInit, handler: () => T): T {
  return withGregaleReleaseContext(new Headers(headers).get(GREGALE_RELEASE_HEADER), handler);
}

/** Return the current request's release, when one was selected. */
export function currentGregaleRelease(): string | undefined {
  return releaseContext.getStore();
}

function urlOf(input: RequestInfo | URL): URL | undefined {
  try {
    if (input instanceof URL) return input;
    if (input instanceof Request) return new URL(input.url);
    const base = typeof location === 'undefined' ? undefined : location.href;
    return new URL(input, base);
  } catch {
    return undefined;
  }
}

function isManagedGregaleService(url: URL): boolean {
  return url.hostname.toLowerCase().replace(/\.$/, '').endsWith('.svc.gregale');
}

/**
 * Create a fetch function that forwards the current release to managed
 * Gregale service hosts only. It strips X-Gregale-Revision on those hops
 * because that pin belongs to the caller app, not the downstream app. Other
 * hosts and their headers are left untouched.
 *
 * Use it once at app startup and wrap inbound handlers with
 * `withGregaleRequestContext(request.headers, handler)`.
 */
export function createGregaleFetch(fetchImpl: typeof globalThis.fetch = globalThis.fetch): typeof globalThis.fetch {
  return async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = urlOf(input);
    if (!url || !isManagedGregaleService(url)) return fetchImpl(input, init);

    const headers = new Headers(input instanceof Request ? input.headers : undefined);
    new Headers(init?.headers).forEach((value, name) => headers.set(name, value));

    // A revision is scoped to the current app and must never escape to a
    // downstream service. The graph release is the only cross-service pin.
    headers.delete(GREGALE_REVISION_HEADER);
    const release = cleanRelease(headers.get(GREGALE_RELEASE_HEADER)) ?? currentGregaleRelease();
    if (release) headers.set(GREGALE_RELEASE_HEADER, release);
    else headers.delete(GREGALE_RELEASE_HEADER);

    return fetchImpl(input, { ...init, headers });
  };
}
